### PR TITLE
Initial version of the dotfiles.

### PR DESCRIPTION
**Pull request by smebberson**
This is a first commit (taken from smebberson/dotfiles) for an Idearium dotfiles. The intent is that all Idearium developers would use this dotfiles to improve developer productivity and to allow a tighter integratiob between Idearium development practices and the command line.

Please have a review of all code. You can also test it on your local to make sure it all works. Yell out if there any glaring omissions or issues you think should stop this from being the first version.

### CODE REVIEW

File: platform/linux/.gitconfig [hunk 1]
```diff
@@ -0,0 +1,9 @@
+[user]
+	name = Scott Mebberson
+	email = scott@scottmebberson.com
```
Comment (allanchau) [inline platform/linux/.gitconfig]: Hmm is there a way we can have this without specifying the user?

File: z.sh [hunk 2]
```diff
@@ -0,0 +1,243 @@
+# Copyright (c) 2009 rupa deadwyler under the WTFPL license
```
Comment (allanchau) [inline z.sh]: This has been updated to a v2 although it seems to work the same.

**Review by allanchau — COMMENTED** [review]
Some discussion points, I'll add more in the comments.

File: platform/linux/.gitconfig (hunk 1 above)
Comment (smebberson) [reply platform/linux/.gitconfig]: Oh right, indeed.

Comment (smebberson) [reply platform/linux/.gitconfig]: I'll have to improve this via the install process. I'll do that shortly.

**Comment by allanchau** [comment]
I haven't used dot file on linux but we should add `.inputrc` on Mac https://github.com/allanchau/dotfiles/blob/master/src/.inputrc

What do you reckon about adding `.bash_prompt`? https://github.com/allanchau/dotfiles/blob/master/src/.bash_prompt
I have mine setup to show `master ✓ ~/Developer/dotfiles $ ` but it is personal preference I guess.

Aliases? https://github.com/allanchau/dotfiles/blob/master/src/.alias

Not everyone uses Homebrew, but do you know of a native way to install + setup apps? https://github.com/allanchau/dotfiles/blob/master/scripts/brew-cask.sh https://github.com/allanchau/dotfiles/blob/master/apps/atom/setup.sh

**Comment by smebberson** [comment]
@allanchau, there were a bunch of missing files :( I've added them back in. Also update the `./install` process to ensure the user can customise the `git config` `user.name` and `user.email` values. Please take another look.

File: .bash_aliases [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+# command shortcuts
+alias lsl='ls -l'
+alias lsa='ls -la'
+
+# vagrant shortcuts
+alias v='vagrant'
+alias vs='vagrant ssh'
+
+# git command shortcuts
+alias gs='git status'
```
Comment (allanchau) [inline .bash_aliases]: `git status -s` is much prettier :)

File: .bash_aliases [hunk 4]
```diff
@@ -0,0 +1,42 @@
+
+# command shortcuts
+alias lsl='ls -l'
+alias lsa='ls -la'
+
+# vagrant shortcuts
+alias v='vagrant'
+alias vs='vagrant ssh'
```
Comment (allanchau) [inline .bash_aliases]: I also have these.
```shell
alias vh="vagrant halt"
alias vu="vagrant up"
```

File: .bash_aliases [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+	git checkout -b "$1" && git push -u origin "$1"
+}
+
+function gtag () {
+	git tag -a "$1" -m "$2" && git push origin "$1"
+}
+
+function gbdo () {
+	git branch -D "$1" && git push origin ":$1"
+}
```
Comment (allanchau) [inline .bash_aliases:53]: These also comes in handy for me, depends how lazy you feel lol
```shell
# dir traversal
alias ..="cd .."
alias ...="cd ../.."
alias ....="cd ../../.."

# Recursively remove Apple meta files
alias cleanupds="find . -type f -name '*.DS_Store' -ls -exec /bin/rm {} \;"
```

**Comment by smebberson** [comment]
@allanchau, just added `.inputrc`.

**Comment by allanchau** [comment]
@smebberson I don't really want to override my existing setup, so do you think it is ok if I don't run the `install` script?

File: platform/linux/.gitconfig [hunk 6]
```diff
@@ -0,0 +1,6 @@
+[credential]
+	helper = cache --timeout=172800
+[diff]
+	external = /home/smebberson/diff.sh
```
Comment (allanchau) [inline platform/linux/.gitconfig]: @smebberson  This one still points to `smebberson`, I think the diff file is missing?

**Comment by smebberson** [comment]
@allanchau, just added those aliases that you have.

File: .bash_aliases (hunk 4 above)
Comment (smebberson) [reply .bash_aliases]: Added

File: .bash_aliases (hunk 3 above)
Comment (smebberson) [reply .bash_aliases]: Didn't know about that. Added :)

File: .bash_aliases (hunk 5 above)
Comment (smebberson) [reply .bash_aliases:53]: Added those.

File: platform/linux/.gitconfig (hunk 6 above)
Comment (smebberson) [reply platform/linux/.gitconfig]: Thanks. Fixed that.

**Comment by smebberson** [comment]
@allanchau, I've update the `.bash_prompt` to be more like yours.

**Comment by smebberson** [comment]
@allanchau, why don't you want to override your setup? Isn't that the point of having these? So that all Idearium devs run the same dotfiles and benefit from updates and improvements to them?

**Comment by allanchau** [comment]
@smebberson Yeah but I have some specific ones that wouldn't be useful for everyone `Homebrew`, `Logitech options` etc. Also I've set mine up to customise atom + Hyper with light themes which most people hate 👎 .

**Comment by smebberson** [comment]
@allanchau, can we allow an avenue for customisation?

**Comment by smebberson** [comment]
@allanchau, the reason I want this is that the devops environment depends on a few settings. So I was thinking this would be a good place to create them so that its super simple to get started as an Idearium developer.

**Comment by allanchau** [comment]
@smebberson It would be good if the install script could run some custom code in the install script `custom.sh`? We can then move any customisations there? I'll run the install now

**Comment by allanchau** [comment]
@smebberson is it possible to customise the dotfile path?

**Comment by smebberson** [comment]
@allanchau, what do you mean by the dotfile path? What are you thinking through?

**Comment by smebberson** [comment]
@allanchau, `custom.sh` is a good idea.

**Comment by allanchau** [comment]
I currently have personal projects in `~/Developer/` and idearium projects in `~/Developer/idearium/`. It helps me scan over projects better but others may want it in `~/Library/Web` etc?

**Comment by smebberson** [comment]
@allanchau, just added support for a `custom.sh`. Check it out and let me know if it does/doesn't suit your needs...

**Comment by smebberson** [comment]
@allanchau, I think I can make it so that this repository can live anywhere. Let me have a crack at it.

**Comment by smebberson** [comment]
@allanchau, is the `custom.sh` good enough?

**Comment by smebberson** [comment]
@allanchau, this repository can now live anywhere and still work :)

**Comment by allanchau** [comment]
@smebberson I am getting a few errors on install:
```shell
/Users/allanchau/.bash_profile: line 14: /Users/allanchau/.bash_colors: No such file or directory
/Users/allanchau/.bash_profile: line 14: /Users/allanchau/.bash_prompt: No such file or directory
/Users/allanchau/.bash_profile: line 14: /Users/allanchau/.bash_exports: No such file or directory
```

**Comment by smebberson** [comment]
@allanchau, I just ran it on @alicecoleman's MacBook Pro with no issue at all. You said you did a search and replace on some paths, do you think that could be the issue? Can you please remove any changes you've made and try again?

**Comment by smebberson** [comment]
@allanchau, I just did a search on the code base, and there is no reference to `Library`, other than the `README.md`. Do you have the latest version?

**Comment by allanchau** [comment]
@smebberson I think I found the problem, `.bash_exports` is missing? I added that and the install worked.

File: .bash_aliases (hunk 5 above)
Comment (allanchau) [reply .bash_aliases:53]: @smebberson One more thing, can we add an alias for git? `g`

**Comment by smebberson** [comment]
@allanchau, okay, I've just made a change that should update this to work better. Could you please try again?

**Comment by allanchau** [comment]
@smebberson What do you reckon about replacing `dt` with the `dtl` alias? I've never actually used `git diff` on it's own.
Also some more aliases for you to consider 
```shell
alias gm='git mergetool' # instead of gmt?
alias gpl='git pull --rebase' # Add --rebase
```

**Comment by smebberson** [comment]
@allanchau, I added `--rebase` to `gpl`. I also added `gm` and `gmt`. I think we need both. I use `gm` and `gmt` from time to time. Like wise with `gd` and `gdt`.

**Comment by allanchau** [comment]
@smebberson Cool, it seems to work good for me now. Happy to merge whenever.

**Pull request merged**